design-proposals/export-import/volume-export.md

```diff
@@ -0,0 +1,254 @@
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: vm1
+```
+
+### Export a VirtualMachineSnapshot
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-snap1
+spec:
+    source:
+        apiGroup: snapshot.kubevirt.io
+        kind: VirtualMachineSnapshot
+        name: snap1
+```
+
+### Export a PersistentVolumeClaim
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-snap1
+spec:
+    source:
+        apiGroup: v1
+        kind: PersistentVolumeClaim
+        name: pvc1
+```
+
+### VirtualMachineExport Status
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    namespace: ns1
+    name: export-vm1
+spec:
+    ...
+status:
+    phase: Ready # Pending|Ready|Terminated
+    token: 8fmf94kf
+    links:
+      internal:
+        cert: <base64 encoded cert>
+        volumes:
+        - name: kubevirt-volume
+          - format: raw
+            url: https://virt-export.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/disk.img?token=8fmf94kf
+          - format: gzip
+            url: https://virt-export.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/disk.img.gz?token=8fmf94kf
+        - name: archive-volume
+          - format: dir
+            url: https://virt-export.kubevirt/ns1/export-vm1/volumes/archive-volume/dir?token=8fmf94kf
+          - format: tar.gz
+            url: https://virt-export.kubevirt/ns1/export-vm1/volumes/archive-volume/disk.tar.gz?token=8fmf94kf
+      external:
+        cert: <base64 encoded cert>
+        volumes:
+        - name: kubevirt-volume
+          - format: raw
+            url: https://virt-export.kubevirt.my-cluster/ns1/export-vm1/volumes/kubevirt-volume/disk.img?token=8fmf94kf
+          - format: gzip
+            url: https://virt-export.kubevirt.my-cluster/ns1/export-vm1/volumes/kubevirt-volume/disk.img.gz?token=8fmf94kf
+        - name: archive-volume
+          - format: dir
+            url: https://virt-export.kubevirt.my-cluster/ns1/export-vm1/volumes/archive-volume/dir?token=8fmf94kf
+          - format: tar.gz
+            url: https://virt-export.kubevirt.my-cluster/ns1/export-vm1/volumes/archive-volume/disk.tar.gz?token=8fmf94kf
+    conditions:
+    - type: Ready
+      status: True
+      ...
+      reason: Ready to export Virtual Machine
+```
+
+* Cluster must have Ingress/Route support for `external` urls and certificate to be set
+
+## virtctl vmexport
+
+virtctl will be extended to include an `vmexport` command that will create/delete a `VirtualMachineExport`
+
+```
+virtctl vmexport create --vm=vm1 vm1-export
+```
+
+```
+virtctl vmexport create --snapshot=snap1 snap1-export
+```
+
+```
+virtctl vmexport delete snap1-export
+```
+
+Once a `VirtualMachineExport` is created, `virtctl vmexport` can download volume archives
+
+```
+virtctl vmexport download vm1-export --volume=volume1 --output disk.img.gz
+```
```

Comment:
the virtctl image-upload command has the ability to both create a new datavolume and upload to it in a single command.

Could we do something similar for export, where someone can download a volume for a VM using a single command?

Reply:
Yeah, would have to somehow configure the source/target cluster kubeconfigs somehow though

Reply:
i just meant the download action. Right now someone has to issue two commands `create` the export and `download` the volume

I was saying, can we streamline this and imply the create during the download, similar to how we imply the pvc create during the upload